File: CloudStorage.Infrastructure/Services/AdminService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using CloudStorage.Application.DTOs;
using CloudStorage.Application.Interfaces;
using CloudStorage.Domain.Entities;
using CloudStorage.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Diagnostics.HealthChecks;
using BCrypt.Net;

namespace CloudStorage.Infrastructure.Services
{
    public class AdminService : IAdminService
    {
        private readonly ApplicationDbContext _context;
        private readonly IRefreshTokenService _refreshTokenService;
        private readonly ILogger<AdminService> _logger;
        private readonly HealthCheckService _healthCheckService;

        public AdminService(
            ApplicationDbContext context,
            IRefreshTokenService refreshTokenService,
            ILogger<AdminService> logger,
            HealthCheckService healthCheckService)
        {
            _context = context;
            _refreshTokenService = refreshTokenService;
            _logger = logger;
            _healthCheckService = healthCheckService;
        }

        public async Task<AdminDashboardStatsDto> GetDashboardStatsAsync()
        {
            _logger.LogInformation("Fetching admin dashboard statistics from real-time data");

            var totalFiles = await _context.FileMetadata.AsNoTracking().CountAsync(f => !f.IsDeleted);
            var totalUsers = await _context.Users.AsNoTracking().CountAsync();
            var suspendedUsers = await _context.Users.AsNoTracking().CountAsync(u => !u.IsActive);
            var totalStorageUsed = await _context.FileMetadata.AsNoTracking().Where(f => !f.IsDeleted).SumAsync(f => f.Size);
            var totalStorageLimit = await _context.Users.AsNoTracking().SumAsync(u => u.StorageQuota);

            var now = DateTime.UtcNow;
            var today = now.Date;
            var last24h = now.AddHours(-24);
            var previous24h = last24h.AddHours(-24);
            var lastWeek = today.AddDays(-7);
            var previousWeek = lastWeek.AddDays(-7);

            // Calculate Trends
            var activeUsersNow = await _context.Users.CountAsync(u => u.LastLoginAt >= last24h);
            var activeUsersPrev = await _context.Users.CountAsync(u => u.LastLoginAt >= previous24h && u.LastLoginAt < last24h);
            var usersTrend = activeUsersPrev == 0 ? 0 : Math.Round(((double)(activeUsersNow - activeUsersPrev) / activeUsersPrev) * 100, 1);

            var uploadsToday = await _context.ActivityLogs.CountAsync(a => a.Timestamp >= today && a.Action == "UPLOAD");
            var downloadsToday = await _context.ActivityLogs.CountAsync(a => a.Timestamp >= today && (a.Action == "DOWNLOAD" || a.Action == "DOWNLOAD_CHUNK"));
            var newUsersThisWeek = await _context.Users.CountAsync(u => u.CreatedAt >= lastWeek);

            // Generate History (Last 7 Days)
            var storageHistory = new List<double>();
            var trafficHistory = new List<double>();
            for (int i = 6; i >= 0; i--)
            {
                var date = today.AddDays(-i);
                var nextDate = date.AddDays(1);

                // For storage, we'd ideally have snapshots. As a proxy, we'll use cumulative size at that point.
                var sizeAtDate = await _context.FileMetadata.AsNoTracking()
                    .Where(f => f.CreatedAt < nextDate && !f.IsDeleted)
                    .SumAsync(f => f.Size);
                storageHistory.Add((double)sizeAtDate / (1024 * 1024 * 1024)); // GB

                var trafficAtDate = await _context.ActivityLogs.AsNoTracking()
                    .CountAsync(a => a.Timestamp >= date && a.Timestamp < nextDate);
                trafficHistory.Add(trafficAtDate);
            }

            // Calculate Regional Telemetry (Simulated mapping from real traffic)
            var totalRecentTraffic = await _context.ActivityLogs.CountAsync(a => a.Timestamp >= last24h);
            var regionalTraffic = new List<RegionalNodeDto>
            {
                new() { Id = "NA_HUB", RegionName = "North America", X = 200, Y = 150, Intensity = CalculateIntensity(totalRecentTraffic, 0.35) },
                new() { Id = "EU_HUB", RegionName = "Europe", X = 500, Y = 150, Intensity = CalculateIntensity(totalRecentTraffic, 0.25) },
                new() { Id = "AS_HUB", RegionName = "Asia", X = 750, Y = 200, Intensity = CalculateIntensity(totalRecentTraffic, 0.20) },
                new() { Id = "SA_HUB", RegionName = "South America", X = 300, Y = 350, Intensity = CalculateIntensity(totalRecentTraffic, 0.10) },
                new() { Id = "AF_HUB", RegionName = "Africa", X = 550, Y = 280, Intensity = CalculateIntensity(totalRecentTraffic, 0.05) },
                new() { Id = "AU_HUB", RegionName = "Australia", X = 800, Y = 380, Intensity = CalculateIntensity(totalRecentTraffic, 0.05) }
            };

            return new AdminDashboardStatsDto
            {
                TotalFiles = totalFiles,
                TotalUsers = totalUsers,
                TotalStorageUsed = totalStorageUsed,
                TotalStorageLimit = totalStorageLimit,
                ActiveUsersLast24h = activeUsersNow,
                SuspendedUsers = suspendedUsers,
                UploadsToday = uploadsToday,
                DownloadsToday = downloadsToday,
                NewUsersThisWeek = newUsersThisWeek,
                ActiveSessionsNow = activeUsersNow,
                FilesTrend = 0, // Would require file snapshot tracking
                UsersTrend = usersTrend,
                StorageTrend = 0,
                StorageHistory = storageHistory,
                TrafficHistory = trafficHistory,
                RegionalTraffic = regionalTraffic
            };
        }

        private double CalculateIntensity(int totalTraffic, double weight)
        {
            // Normalize intensity based on traffic volume (0.2 to 1.0 range)
            if (totalTraffic == 0) return 0.1;
            var baseIntensity = Math.Min(1.0, 0.2 + (totalTraffic / 1000.0));
            return Math.Round(baseIntensity * weight * 2, 2); // Boost for visibility
        }

        public async Task<IEnumerable<AdminUserManagementDto>> GetAllUsersAsync()
        {
            _logger.LogInformation("Retrieving all users for admin management");
            return await _context.Users
                .AsNoTracking()
                .OrderByDescending(u => u.CreatedAt)
                .Select(u => new AdminUserManagementDto
                {
                    Id = u.Id,
                    Username = u.Username,
                    Email = u.Email,
                    Role = u.Role,
                    IsActive = u.IsActive,
                    CreatedAt = u.CreatedAt,
                    LastLoginAt = u.LastLoginAt,
                    StorageUsed = _context.FileMetadata.Where(f => f.OwnerId == u.Id && !f.IsDeleted).Sum(f => f.Size),
                    StorageQuota = u.StorageQuota
                })
                .ToListAsync();
        }

        public async Task<AdminUserManagementDto?> UpdateUserQuotaAsync(int userId, long newQuota)
        {
            _logger.LogInformation("Adjusting storage quota for user {UserId} to {NewQuota}", userId, newQuota);
            var user = await _context.Users.FindAsync(userId);
            if (user == null) return null;

            user.StorageQuota = newQuota;
            await _context.SaveChangesAsync();

            return await GetUserManagementDetailsAsync(userId);
        }

        public async Task<SystemHealthDetailsDto> GetSystemHealthAsync()
        {
            _logger.LogInformation("Performing real system health diagnostic check");

            var report = await _healthCheckService.CheckHealthAsync();
            var isHealthy = report.Status == HealthStatus.Healthy;

            // Use real environment metrics where possible
            var process = System.Diagnostics.Process.GetCurrentProcess();
            var drive = new System.IO.DriveInfo(System.IO.Path.GetPathRoot(AppDomain.CurrentDomain.BaseDirectory)!);

            return new SystemHealthDetailsDto
            {
                Status = report.Status.ToString().ToLower(),
                CpuUsage = 0, // Hard to get accurately across OS without external libs
                MemoryUsed = process.WorkingSet64,
                MemoryTotal = GC.GetGCMemoryInfo().TotalAvailableMemoryBytes,
                DiskUsed = drive.TotalSize - drive.AvailableFreeSpace,
                DiskTotal = drive.TotalSize,
                Uptime = (int)(DateTime.UtcNow - process.StartTime.ToUniversalTime()).TotalSeconds,
                ActiveConnections = _context.Users.Count(u => u.LastLoginAt >= DateTime.UtcNow.AddMinutes(-5)),
                RequestsPerMinute = _context.ActivityLogs.Count(a => a.Timestamp >= DateTime.UtcNow.AddMinutes(-1)),
                ErrorRate = isHealthy ? 0.01 : 0.15,
                AvgResponseMs = 45, // Placeholder for real middleware telemetry
                Checks = report.Entries.Select(e => new ServiceCheckDto
                {
                    Name = e.Key,
                    Status = e.Value.Status == HealthStatus.Healthy ? "ok" : "unhealthy",
                    Message = e.Value.Description ?? e.Value.Exception?.Message ?? "No description available",
                    LatencyMs = (int)e.Value.Duration.TotalMilliseconds
                }).ToList()
            };
        }

        public async Task<IEnumerable<AdminAuditDto>> GetRecentAuditLogsAsync(int count = 50)
        {
            _logger.LogInformation("Retrieving {Count} recent audit logs", count);
            return await _context.ActivityLogs
                .AsNoTracking()
                .OrderByDescending(a => a.Timestamp)
                .Take(count)
                .Select(a => new AdminAuditDto
                {
                    Id = a.Id,
                    UserId = a.UserId,
                    PerformedBy = a.User.Username,
                    Action = a.Action,
                    TargetType = a.EntityType,
                    TargetName = a.EntityId, // In real app, we might join to get the name
                    Details = a.Details,
                    PerformedAt = a.Timestamp,
                    IpAddress = "127.0.0.1" // Mocked IP for now
                })
                .ToListAsync();
        }

        public async Task<AdminUserManagementDto> CreateUserAsync(CreateUserDto dto)
        {
            _logger.LogInformation("Creating new user with email {Email} and role {Role}", dto.Email, dto.Role);
            if (await _context.Users.AnyAsync(u => u.Email == dto.Email))
                throw new Exception("Email already exists");

            var user = new User
            {
                Username = dto.Username,
                Email = dto.Email,
                Role = dto.Role,
                IsActive = true,
                CreatedAt = DateTime.UtcNow,
                PasswordHash = BCrypt.Net.BCrypt.HashPassword(dto.Password)
            };

            _context.Users.Add(user);
            await _context.SaveChangesAsync();

            return (await GetUserManagementDetailsAsync(user.Id))!;
        }

        public async Task ToggleUserStatusAsync(int userId, bool isActive)
        {
            _logger.LogInformation("Toggling status for user {UserId} to {Status}", userId, isActive ? "Active" : "Inactive");

            await _context.Users
                .Where(u => u.Id == userId)
                .ExecuteUpdateAsync(s => s.SetProperty(u => u.IsActive, isActive));
        }

        public async Task<string> GetImpersonationTokenAsync(int userId)
        {
            // In a production app, this would involve a specialized security handshake
            return "mock_impersonation_token_" + Guid.NewGuid().ToString("N");
        }

        private async Task<AdminUserManagementDto?> GetUserManagementDetailsAsync(int userId)
        {
            return await _context.Users
               .AsNoTracking()
               .Where(u => u.Id == userId)
               .Select(u => new AdminUserManagementDto
               {
                   Id = u.Id,
                   Username = u.Username,
                   Email = u.Email,
                   Role = u.Role,
                   IsActive = u.IsActive,
                   CreatedAt = u.CreatedAt,
                   LastLoginAt = u.LastLoginAt,
                   StorageUsed = _context.FileMetadata.Where(f => f.OwnerId == u.Id && !f.IsDeleted).Sum(f => f.Size),
                   StorageQuota = 5368709120
               })
               .FirstOrDefaultAsync();
        }
    }
}

File: CloudStorage.Client/src/app/core/services/admin.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { HttpParams } from '@angular/common/http';
import { ApiService } from './api.service';
import { BaseService } from '../models/base-service';
import { ApiResponse } from '../models/api-response.model';
import { AuditEntry } from '../models/audit.model';

// ─── Admin Domain Models ────────────────────────────────────────

export interface AdminUser {
  id: number;
  username: string;
  email: string;
  role: string;
  isActive: boolean;
  storageUsed: number;
  storageQuota: number;
  createdAt: string;
  lastLoginAt: string | null;
}

export interface SystemHealth {
  status: 'healthy' | 'degraded' | 'critical';
  uptime: number; // seconds
  cpuUsage: number; // 0-100
  memoryUsed: number; // bytes
  memoryTotal: number; // bytes
  diskUsed: number; // bytes
  diskTotal: number; // bytes
  activeConnections: number;
  requestsPerMinute: number;
  errorRate: number; // 0-100
  avgResponseMs: number;
  checks: HealthCheck[];
}

export interface HealthCheck {
  name: string;
  status: 'ok' | 'warn' | 'error';
  message: string;
  latencyMs?: number;
}

export interface AdminStats {
  totalFiles: number;
  totalUsers: number;
  totalStorageUsed: number;
  totalStorageLimit: number;
  activeUsersLast24h: number;
  suspendedUsers: number;
  uploadsToday: number;
  downloadsToday: number;
  newUsersThisWeek: number;
  activeSessionsNow: number;
  filesTrend: number;
  usersTrend: number;
  storageTrend: number;
  storageHistory: number[];
  trafficHistory: number[];
}

export interface AdminAuditFilter {
  count?: number;
}

// ─── Admin Service ──────────────────────────────────────────────

@Injectable({ providedIn: 'root' })
export class AdminService extends BaseService {
  private api = inject(ApiService);

  // ── User Management ──────────────────────────────────────────

  public getUsers(): Observable<AdminUser[]> {
    return this.api.get<ApiResponse<AdminUser[]>>('/admin/users').pipe(
      map(r => r.data ?? []),
      catchError(this.handleError<AdminUser[]>('GET_ADMIN_USERS', []))
    );
  }

  public createUser(user: Partial<AdminUser>): Observable<AdminUser> {
    const payload = {
      username: user.username,
      email: user.email,
      password: (user as any).password, // Explicitly pass password from form
      role: user.role,
      initialQuota: user.storageQuota
    };
    return this.api.post<ApiResponse<AdminUser>>('/admin/users', payload).pipe(
      map(r => r.data!),
      catchError(this.handleError<AdminUser>('CREATE_USER'))
    );
  }

  public updateUserQuota(id: number, newQuota: number): Observable<AdminUser> {
    return this.api.patch<ApiResponse<AdminUser>>(`/admin/users/${id}/quota`, { newQuota }).pipe(
      map(r => r.data!),
      catchError(this.handleError<AdminUser>('UPDATE_QUOTA'))
    );
  }

  public toggleUserStatus(id: number, isActive: boolean): Observable<void> {
    return this.api.post<ApiResponse<void>>(`/admin/users/${id}/toggle-status`, isActive).pipe(
      map(() => void 0),
      catchError(this.handleError<void>('TOGGLE_USER_STATUS'))
    );
  }

  public impersonateUser(id: number): Observable<{ token: string }> {
    return this.api.post<ApiResponse<{ token: string }>>(`/admin/users/${id}/impersonate`, {}).pipe(
      map(r => r.data!),
      catchError(this.handleError<{ token: string }>('IMPERSONATE_USER'))
    );
  }

  // ── System Health ─────────────────────────────────────────────

  public getSystemHealth(): Observable<SystemHealth> {
    return this.api.get<ApiResponse<SystemHealth>>('/admin/health').pipe(
      map(r => r.data!),
      catchError(this.handleError<SystemHealth>('GET_SYSTEM_HEALTH'))
    );
  }

  public getAdminStats(): Observable<AdminStats> {
    return this.api.get<ApiResponse<AdminStats>>('/admin/stats').pipe(
      map(r => r.data!),
      catchError(this.handleError<AdminStats>('GET_ADMIN_STATS'))
    );
  }

  // ── Audit Logs ────────────────────────────────────────────────

  public getAuditLogs(count = 50): Observable<AuditEntry[]> {
    const params = new HttpParams().set('count', count.toString());

    return this.api.get<ApiResponse<AuditEntry[]>>('/admin/audit', params).pipe(
      map(r => r.data ?? []),
      catchError(this.handleError<AuditEntry[]>('GET_ADMIN_AUDIT_LOGS', []))
    );
  }
}
